WooCommerce/Classes/CardPresentPayments/CardPaymentsAdaptor.swift

```diff
@@ -0,0 +1,147 @@
+import Foundation
+import Yosemite
+import class WooFoundation.CurrencyFormatter
+
+enum CardPresentPaymentResult {
+    case success(Order)
+    case failure(CardPaymentErrorProtocol)
+    case cancellation
+}
+
+class CardPresentPaymentsAdaptor {
+    private let currencyFormatter: CurrencyFormatter = CurrencyFormatter(currencySettings: ServiceLocator.currencySettings)
+    private let siteID: Int64
+
+    // These connection controllers shouldn't be lazy, I'm just being lazy.
+    // They're lazy because of `self` implementing `alertsPresenter`.
+    // If that were done by another owned struct, and the handler passed to it as and when it changed, we could do this in the init.
+    private lazy var bluetoothConnectionController: CardReaderConnectionController = {
+        CardReaderConnectionController(
+            forSiteID: siteID,
+            knownReaderProvider: CardReaderSettingsKnownReaderStorage(),
+            alertsPresenter: self,
+            alertsProvider: BluetoothReaderConnectionAlertsProvider(),
+            configuration: CardPresentConfigurationLoader().configuration,
+            analyticsTracker: CardReaderConnectionAnalyticsTracker(
+                configuration: CardPresentConfigurationLoader().configuration,
+                siteID: siteID,
+                connectionType: .userInitiated, // This will need a tweak – it changes from use to use.
+                analytics: ServiceLocator.analytics))
+    }()
+
+    private lazy var tapToPayConnectionController: BuiltInCardReaderConnectionController = {
+        BuiltInCardReaderConnectionController(
+            forSiteID: siteID,
+            alertsPresenter: self,
+            alertsProvider: BluetoothReaderConnectionAlertsProvider(),
+            configuration: CardPresentConfigurationLoader().configuration,
+            analyticsTracker: CardReaderConnectionAnalyticsTracker(
+                configuration: CardPresentConfigurationLoader().configuration,
+                siteID: siteID,
+                connectionType: .userInitiated, // This will need a tweak – it changes from use to use.
+                analytics: ServiceLocator.analytics))
+    }()
+
+    var paymentsAlertHandler: CardPresentPaymentsAlertHandling?
+
+    init(siteID: Int64) {
+        self.siteID = siteID
+    }
+
+    func collectPayment(for order: Order,
+                        using discoveryMethod: CardReaderDiscoveryMethod) async -> CardPresentPaymentResult {
+        let preflightController = await CardPresentPaymentPreflightController(
+            siteID: siteID,
+            configuration: CardPresentConfigurationLoader().configuration,
+            rootViewController: UIViewController(),
```

Comment:
just for my learning, it looks like we can just pass any view controller here and in `CollectOrderPaymentUseCase`, does that mean this `rootViewController` parameter isn't really necessary?

Reply:
At the moment it is used, by the existing code. I'm only passing this as a temporary measure so that I could work on onboarding, etc. Here's the list where it's used at the moment:

- `CardPresentPaymentAlertsPresenter`
	- `present(present, animated: false)` in `dismissSeveralFoundAndPresent` and `dismissCommonAndPresent` 
	- these are used to show the UIKit payment alerts, so in our adapted case we won't need to use them because we'll publish the view models for the caller to display those alerts.
- `CollectOrderPaymentUseCase`
	- `self.rootViewController.presentedViewController?.dismiss()` in failure handler for `attemptPayment` in `collectPayment`.
	    - I think that by managing our own display of alerts in the caller, we can avoid this need too
	- `present` used in receipts
	    - Receipts – I'm not tackling these head on at the moment, because they're really something that comes after the payment flow.
	- set as `noticePresenter.presentingViewController` when receipt printing fails
- `CardPresentPreflightController`
	- `onboardingPresenter.showOnboardingIfRequired(from: rootViewController)` – 
	    - safe to ignore since we don't use it in our implementation of the adaptor for onboarding

So all in all, from the adaptor we may (end up) being able to get away without a VC here. 

I'd do that by replacing the requirement for a UIViewController, with a requirement for a protocol with the above methods, and then pass in a null object from the adaptor so we don't end up with warnings or UIKit getting confused about a VC outside the view hierarchy being used for presentation.